EVS Framework: Support NemotronH_Nano_VL_V2

netanel-haber opened the pull request:
<!-- Thank you for your contribution! Please follow these guidelines to enhance your pull request. If anything is unclear, submit your PR and reach out to maintainers for assistance. Join our Slack community at https://slack.sglang.ai to discuss further. -->

## Motivation

> https://arxiv.org/abs/2510.14624: Efficient Video Sampling: Pruning Temporally Redundant Tokens for Faster VLM Inference

Add support for EVS pruning to `NemotronH_Nano_VL_V2`. See `python/sglang/srt/multimodal/evs/README.md` for more details.

A pruning rate of 0.7, for example, means that 30% of the vision encoder output embeddings will be *retained*, and added to the language model prompt.

Nemotron-Nano-VL-V2.0 was trained with a pruning rate of `0.7`, which is the default value in the model config, so EVS will be on by default. To disable or otherwise configure the pruning rate (accuracy may vary), set e.g.: `--json-model-override-args '{"video_pruning_rate": 0.0}'` [EVS off].

Example of full serving command with disabled EVS:

```bash
sglang serve --model-path nvidia/Nemotron-Nano-12B-v2-VL-BF16 --trust-remote-code --json-model-override-args '{"video_pruning_rate": 0.0}'
```

The PR creates a framework that should make it easier to add EVS support to additional models in the future, such as Qwen2.5VL, which will be implemented in a following pr. 

## Benchmarks

NanoV2VLM on SGLang and VLLM, with `video_pruning_rate=0.9` got an identical overall VideoMME score of `0.644`.

| Scenario\Metric                	        | Online TTFT (Seconds) stderr: ±0.38 	| VideoMME Accuracy       	|
|---------------------------------	        |-------------------------------------	                |-------------------------	                |
| EVS Off [q=0.0]                 	        | 11.96 [100%]                        	                | Between 0.665 and 0.668 	|
| EVS Off [q=0.4]                 	        | 09.97 [ 83%]                        	                |                         	                |
| EVS On  [q=0.7] (default value) 	| 08.79 [ 73%]                        	                |                         	                |
| EVS Off [q=0.9]                 	        | 08.39 [ 70%]                        	                | 0.644                   	                |

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.ai/developer_guide/contribution_guide.html#code-style-guidance).
- [x] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

netanel-haber committed:
[EVS] Support NemotronH_Nano_VL_V2

netanel-haber committed:
Fix chunked prefill EVS

yuan-luo commented:
@netanel-haber Could you provide the TTFT benchmark diagram for this feature on different pruning rates? So as the user can have better decision in the balance of performance and accuracy.

netanel-haber committed:
prefix caching: always use fresh pre_chunking_input_ids, don't save it on EVSEmbeddingResult, since it might be from mm_cache and therefore from a previous prompt

netanel-haber committed:
add perf metrics

netanel-haber committed:
Merge branch 'main' into support-evs-for-nemotronh-nano-vl-v2

yhyang201 committed:
Merge branch 'main' into support-evs-for-nemotronh-nano-vl-v2

netanel-haber committed:
Merge branch 'main' into support-evs-for-nemotronh-nano-vl-v2

yhyang201 commented:
/rerun-failed-ci

yhyang201 committed:
Merge branch 'main' into support-evs-for-nemotronh-nano-vl-v2

JustinTong0323 commented on test/srt/test_evs.py:
This test only includes the config check. Should we add additional tests for EVS's functionality?

netanel-haber committed:
Merge branch 'main' into support-evs-for-nemotronh-nano-vl-v2

netanel-haber commented on python/sglang/srt/mem_cache/multimodal_cache.py:
```diff
@@ -94,17 +100,20 @@ def get(
-        self, mm_hash: int, embedding: torch.Tensor, loc: Optional[torch.Tensor] = None
+        self,
+        mm_hash: int,
+        embedding: EmbeddingResult,
```
The reason I did this is because I need to tokens_per_frame that is created along with the embedding in the case of EVS - so if the embedding is cached, I need tokens per frame cached as well. The way I implemented this change, the cache is agnostic to whatever else I save on an entry, so it needn't be aware of EVS etc.

netanel-haber committed:
- Changed the way embeddings are retrieved and stored in MMEncoder to use the new EmbeddingResult structure.
- Added an assertion in MultiModalStaticCache to ensure that the embedding being set is of type EmbeddingResult.
